### PR TITLE
Fixes integer overflow for efficient filtering

### LINKED ISSUE #2901: [BUG] Integer overflow for MDC calculations
**What is the bug?**

When there is a filter query present in KNN, the efficient filtering algorithm tries to estimate distance computation by multiplying query dimensions and filtered doc count. This multiplication causes integer overflow if filtered doc counts are high. 

This ultimately forces an exact KNN search behavior where ANN search is required causing higher latencies 

**What is the expected behavior?**
ANN should be executed when filterid count is large

**What is your host/environment?**
 - OS:linux
 - Version 2.19
 - Plugins

### PR DESCRIPTION
**Pull request by shatejas**
Estimating distance calculation can run into int overflow if the filter docs cardinality is high. This happens because distance calculation involves multiplying dimension and filter docs cardinality

### Related Issues
Resolves [#[2901]](https://github.com/opensearch-project/k-NN/issues/2901)

### Check List
- [x] New functionality includes testing.
- [x] Commits are signed per the DCO using `--signoff`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by shatejas** [comment]
Will fix the CIs if they fail

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java [hunk 1]
```diff
@@ -170,6 +182,10 @@ public KNNQuery filterQuery(Query filterQuery) {
         return this;
     }
 
+    public int getQueryDimension() {
+        return this.queryVector != null ? this.queryVector.length : this.byteQueryVector.length;
+    }
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/index/query/KNNQuery.java:196]: for binary vectors the dimension will be dimension*8.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 2]
```diff
@@ -182,14 +182,14 @@ private StringBuilder getLeafLevelExplanation(LeafReaderContext context) {
...
                     .append(cardinality);
-            } else if (!isExactSearchThresholdSettingSet(filterThresholdValue) && isMDCGreaterThanFilterIdCnt(cardinality)) {
+            } else if (!isExactSearchThresholdSettingSet(filterThresholdValue) && isMaxDistCompGreaterThanEstimatedDistComp(cardinality)) {
                 sb.append(KNNConstants.EXACT_SEARCH)
                     .append(" since max distance computation = ")
                     .append(KNNConstants.MAX_DISTANCE_COMPUTATIONS)
-                    .append(" is greater than or equal to cardinality = ")
-                    .append(cardinality);
+                    .append(" is greater than or equal to estimated distance computations = ")
+                    .append(((long) cardinality) * knnQuery.getQueryDimension());
```
Comment (Vikasht34) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:194]: Will String will be trucated with explain API , Can we add test like filtercadinatility = 2^31 and Dimesion 10^3 to verify we don't truncate the string.

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java [hunk 3]
```diff
@@ -68,6 +68,10 @@ public class KNNQuery extends Query {
     // TODO: ThreadContext does not work with logger, remove this from here once its figured out
     private int shardId;
 
+    /**
+     * @deprecated Use builder instead
+     */
+    @Deprecated
```
Comment (Vikasht34) [inline src/main/java/org/opensearch/knn/index/query/KNNQuery.java:74]: Is this all part of Int Overflow issue?

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java [hunk 4]
```diff
@@ -170,6 +182,10 @@ public KNNQuery filterQuery(Query filterQuery) {
         return this;
     }
 
+    public int getQueryDimension() {
```
Comment (Vikasht34) [inline src/main/java/org/opensearch/knn/index/query/KNNQuery.java:185]: Since we are deprecating couple of Query Constructor, to safe guard method, Since we are doing query.length , Can we please add NPE check on Query?

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 5]
```diff
@@ -593,13 +593,12 @@ private boolean isFilteredExactSearchPreferred(final int filterIdsCount) {
...
          */
-        return isMDCGreaterThanFilterIdCnt(filterIdsCount);
+        return isMaxDistCompGreaterThanEstimatedDistComp(filterIdsCount);
     }
 
-    private boolean isMDCGreaterThanFilterIdCnt(int filterIdsCount) {
-        return KNNConstants.MAX_DISTANCE_COMPUTATIONS >= filterIdsCount * (knnQuery.getQueryVector() != null
-            ? knnQuery.getQueryVector().length
-            : knnQuery.getByteQueryVector().length);
+    private boolean isMaxDistCompGreaterThanEstimatedDistComp(int filterDocsCardinality) {
```
Comment (Vikasht34) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:601]: Should we add one test with Filtercadinality with Integer.max with Dim < 10 that can return correct boolean value. 

another test case would be filtercardinality* dim close to Long.Max_Value and check if estimates handle without warping negative.

Another test case what I was thinking , if we set Filtercardinality to max Value with annFilter null and if full estimation is used without overlflow.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 6]
```diff
@@ -593,13 +593,12 @@ private boolean isFilteredExactSearchPreferred(final int filterIdsCount) {
...
+        long estimatedDistanceComputations = ((long) filterDocsCardinality) * knnQuery.getQueryDimension();
```
Comment (Vikasht34) [inline src/main/java/org/opensearch/knn/index/query/KNNWeight.java:602]: Will it be any issue if mutiple threads paraller running searchleaf with this estimator?

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 5 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:601]: > Should we add one test with Filtercadinality with Integer.max with Dim < 10 that can return correct boolean value.

There is already a unit test added for int overflow, are you asking for something else here? its a private function and I would want to keep it that way since its tested

> another test case would be filtercardinality* dim close to Long.Max_Value and check if estimates handle without warping negative

For OS to hit `Long.MAX_VALUE` it needs to have 92 trillion vectors with 10K dimensions (and more if the dimensions are less) in a segment which is not realistic, so this should be good for now

> Another test case what I was thinking , if we set Filtercardinality to max Value with annFilter null and if full estimation is used without overlflow.

If filter cardinality is `Integer.MAX_VALUE` that means the segment has more than 2 billion docs which does not seem to be realistic.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 6 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:602]: yes its thread-safe. filterDocsCardinality is local to the thread

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 2 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:194]: filtercadinatility = 2^31 is unrealistic since this is more than 2 billion docs in a segment. Is there any evidence that explain strings are being truncated, can you clarify?

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java (hunk 3 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/KNNQuery.java:74]: Nope, just a general cleanup

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java (hunk 4 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/KNNQuery.java:185]: Not the right place for NPE check, I added an assert

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java [hunk 7]
```diff
@@ -170,6 +182,20 @@ public KNNQuery filterQuery(Query filterQuery) {
...
+        return switch (vectorDataType) {
+            case FLOAT, BYTE -> {
+                assert queryVector != null;
+                yield queryVector.length;
+            }
+            case BINARY -> {
+                assert byteQueryVector != null;
+                yield byteQueryVector.length * Byte.SIZE;
+            }
+            default -> queryVector.length;
```
Comment (shatejas) [inline src/main/java/org/opensearch/knn/index/query/KNNQuery.java]: I did not find any `vectorDataType != null` check so keeping a default to be safe

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 2 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:194]: Though String builder supports ~2B Character, 
Since we are appending (long) cardinality) * knnQuery.getQueryDimension() this , Can we verify in test with Max Values which is realistic with Max Dimension Supported  to verify string are not concatenated. ?
I guess in Knn weight test , we already testing with integer.MaxValue , Can we have add similar test here as well to verify the string being returned ?

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java (hunk 4 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/KNNQuery.java:185]: assert queryVector != null is NPE Check , it bubbles up exception upwards.  I am good with assert too.

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java (hunk 7 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/KNNQuery.java]: Thanks , I guess in Default also should be put Assert?

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 5 above)
Comment (Vikasht34) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:601]: I see test contains "when(reader.maxDoc()).thenReturn(Integer.MAX_VALUE) ~2.1 billion , should be fine for above mentioned cases.

**Review by Vikasht34 — DISMISSED** [review]
Overall looks good , some nit picks with test cases.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 2 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:194]: There is an [assert](https://github.com/opensearch-project/k-NN/blob/main/src/test/java/org/opensearch/knn/index/query/ExplainTests.java#L565) on explain. I agree its a `contains` check and not `equals` and it should be tightened but its a bit orthogonal and should be tackled as a improvement on tests

If the concern is `(long) cardinality) * knnQuery.getQueryDimension()` will increase the characters in String; a quintillion would only add about 19 characters, which won't breach the String limits

File: src/main/java/org/opensearch/knn/index/query/KNNQuery.java (hunk 7 above)
Comment (shatejas) [reply src/main/java/org/opensearch/knn/index/query/KNNQuery.java]: added

**Review by Vikasht34 — APPROVED** [review]
Looks Good

**Review by navneet1v — APPROVED** [review]
